src/operator/time_series/radix_tree/updater.rs

```diff
@@ -0,0 +1,406 @@
+/// values associated with individual timestamps.
+///
+/// When done with updates, the client calls `finish` to produce a
+/// sorted vector of `TreeNodeUpdate`s.  They can then convert these
+/// updates into a `Batch` and append it to the tree trace to transform
+/// the tree.
+///
+/// For efficiency, `TreeUpdater` requires that `update_timestamp` is invoked
+/// for monotonically increasing timestamps, so that it only needs to scan
+/// the underlying cursor once without backtracking.
+struct TreeUpdater<'a, TS, A, R, S, TC>
+where
+    S: Semigroup<A>,
+{
+    /// Tree cursor.  `TreeUpdater` performs updates of monotonically
+    /// increasing timestamps; therefore the cursor only moves forward.
+    tree_cursor: TC,
+    /// Tracks current location in the tree.
+    stack: Vec<StackFrame>,
+    /// Accumulated tree updates.
+    updates: &'a mut Vec<TreeNodeUpdate<TS, A>>,
+    phantom: PhantomData<(R, S)>,
+}
+
+impl<'a, 'b, TS, A, R, S, TC> TreeUpdater<'a, TS, A, R, S, TC>
+where
+    TS: PrimInt + Debug,
+    A: Clone + Default + Debug,
+    S: Semigroup<A>,
+    TC: RadixTreeCursor<'b, TS, A, R>,
+    R: HasZero,
+{
+    /// Create a new `TreeUpdater` backed by `tree_cursor`.
+    ///
+    /// # Arguments
+    ///
+    /// * `tree_cursor` must point to the root of the tree.
+    /// * `updates` - empty vector of `TreeNodeUpdate`s.  After `finish` call,
+    ///   this vector will contain sorted tree updates. We require the client to
+    ///   pass the vector for allocation reuse.
+    fn new(tree_cursor: TC, updates: &'a mut Vec<TreeNodeUpdate<TS, A>>) -> Self {
+        debug_assert!(updates.is_empty());
+
+        let mut res = Self {
+            tree_cursor,
+            stack: Vec::with_capacity((size_of::<TS>() * 8) / RADIX_BITS as usize),
+            updates,
+            phantom: PhantomData,
+        };
+
+        if res.tree_cursor.key_valid() {
+            res.tree_cursor.skip_zero_weights();
+            if res.tree_cursor.val_valid() {
+                debug_assert_eq!(res.tree_cursor.key(), &Prefix::full_range());
+                res.push_existing(Prefix::full_range(), res.tree_cursor.val().clone(), 0)
+            } else {
+                // No root node -- tree is empty.  Create new root.
+                res.push_new(Prefix::full_range(), TreeNode::new(), 0);
+            }
+        } else {
+            // No root node -- tree is empty.  Create new root.
+            res.push_new(Prefix::full_range(), TreeNode::new(), 0);
+        };
+        res
+    }
+
+    fn finish(mut self) {
+        while !self.stack.is_empty() {
+            self.pop();
+        }
+
+        // TODO: This can be expensive.  Can we keep `updates` ordered so no sorting is
+        // required?  Currently out-of-order updates are added when creating
+        // intermediate tree nodes (`push_new`).  I can think of two tricks to
+        // reduce the cost of sorting: (1) leave gaps in the updates vector when
```

Comment:
there are some clever algorithms that perform insertion sort using gaps. Rob talked about some of them.

Reply:
I never see lists used in Rust

Reply:
> there are some clever algorithms that perform insertion sort using gaps. Rob talked about some of them.

yes, that's the way to do.  I didn't find a good library for it and didn't want to optimize prematurely.

> I never see lists used in Rust

malloc's are expensive.